internal/storage/chunk_uploader_test.go

```diff
@@ -0,0 +1,330 @@
+	contentSize := len(content) // > chunkSize
+	numChunks := int(math.Ceil(float64(contentSize) / float64(chunkSize)))
+
+	AssertGt(contentSize, chunkSize)
+
+	var numCallbacks int
+	var numBytesUploaded int64
+
+	uploader, err := NewChunkUploader(context.Background(), obj, t.req, chunkSize,
+		func(n int64) {
+			numBytesUploaded = n
+			numCallbacks++
+		})
+
+	AssertNe(nil, uploader)
+	AssertEq(nil, err)
+
+	err = uploader.UploadChunkAsync(strings.NewReader(content))
+	AssertEq(nil, err)
+
+	AssertEq(Writing, uploader.state)
+
+	// time.Sleep(1000 * time.Millisecond)
+	if numCallbacks > 0 {
+		AssertGe(numChunks, numCallbacks)
+		AssertGe(contentSize, numBytesUploaded)
+	}
+	AssertEq(numBytesUploaded, uploader.BytesUploadedSoFar())
+	AssertEq(contentSize, uploader.bytesUploadInitiatedSoFar())
+}
+
+func (t *ChunkUploaderTest) TestUploadMultipleMultichunkUploads() {
+	obj := t.obj
+	chunkSize := googleapi.MinUploadChunkSize
+	uploadSizeFactors := [5]float64{1.243, 1.32536, 1.0, 1.63464, 1.99}
+	var contents []string
+	var totalContentSize int
+	var numChunks int
+	for _, factor := range uploadSizeFactors {
+		b := make([]byte, int(factor*float64(chunkSize)))
+		_, err := rand.Read(b)
+		AssertEq(nil, err)
+		contents = append(contents, string(b))
+		contentSize := len(b)
+		totalContentSize += contentSize
+		numChunks += int(math.Ceil(float64(contentSize) / float64(chunkSize)))
+	}
+
+	var numCallbacks int
+	var numBytesUploaded int64
+	uploader, err := NewChunkUploader(context.Background(), obj, t.req, chunkSize,
+		func(n int64) {
+			numBytesUploaded = n
+			numCallbacks++
+		})
+	AssertNe(nil, uploader)
+	AssertEq(nil, err)
+
+	for _, content := range contents {
+		err = uploader.UploadChunkAsync(strings.NewReader(content))
+		AssertEq(nil, err)
+	}
+
+	AssertEq(Writing, uploader.state)
+
+	// time.Sleep(1000 * time.Millisecond)
+	if numCallbacks > 0 {
+		AssertGe(numChunks, numCallbacks)
+		AssertGe(totalContentSize, numBytesUploaded)
+	}
+	AssertEq(numBytesUploaded, uploader.BytesUploadedSoFar())
+	AssertEq(totalContentSize, uploader.bytesUploadInitiatedSoFar())
+}
+
+func (t *ChunkUploaderTest) TestCloseWithoutWrite() {
+	obj := t.obj
+
+	uploader, err := NewChunkUploader(context.Background(), obj, t.req, 1000, nil)
+	AssertNe(nil, uploader)
+	AssertEq(nil, err)
+
+	o, err := uploader.Close()
+	AssertNe(nil, o)
+	AssertEq(nil, err)
+	AssertEq(o.Name, "test_object")
+	AssertEq(0, o.Size)
+}
+
+func (t *ChunkUploaderTest) TestCloseWithSingleUpload() {
+	obj := t.obj
+	chunkSize := 1000
+	contentSize := len(t.content)
+
+	uploader, err := NewChunkUploader(context.Background(), obj, t.req, chunkSize, nil)
+	AssertNe(nil, uploader)
+	AssertEq(nil, err)
+
+	err = uploader.UploadChunkAsync(t.req.Contents)
+	AssertEq(nil, err)
+
+	o, err := uploader.Close()
+	AssertNe(nil, o)
+	AssertEq(nil, err)
+	AssertEq(o.Name, "test_object")
+	AssertEq(contentSize, o.Size)
+}
+
+func (t *ChunkUploaderTest) TestCloseWithMultipleSmallUploads() {
```

Comment:
can you define what does small means in the test method itself.

Reply:
Small stands for where content-size < chunkSize.